### ISSUE TITLE
 CSP error when "name" is displayed on a series chart's tooltipText

### ISSUE THREAD

**Issue by mujahidfa** [amcharts/amcharts4#3578]
### Summary
CSP error when hovering over a series chart with "name" displayed on the tooltipText despite having `options.nonce` set.

### Detailed explanation
Hovering over a `LineSeries` chart that displays a `name` on a series tooltip like so:
```js
let series = chart.series.push(new am4charts.LineSeries());
series.dataFields.valueY = "value";
series.dataFields.dateX = "date";

series.name = "Name";
series.tooltipText = `{name}: [bold]{valueY}[/]`;

chart.cursor = new am4charts.XYCursor();
```
will cause the CSP error as below (on `Label.ts`):

![image](https://user-images.githubusercontent.com/17759705/121460698-f88fdd80-c9df-11eb-883b-77d763496bb0.png)

despite having set the `options.nonce`:
```js
am4core.options.nonce = "eaDHPClCs73k19urySogiQ==";
```

## Reproduction
It's hard to simulate this in CodePen since it requires simulating the `Content-Security-Policy` in a meta tag, which does not work well with CodePen. The best way to simulate this is to copy/paste the below to an `index.html` and open in a browser. Then, hover over the line chart to trigger the tooltips and observe the CSP error in the Devtools:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta
      http-equiv="Content-Security-Policy"
      content="default-src 'self'; img-src 'self' blob: data:; connect-src 'self';
      upgrade-insecure-requests; style-src 'self' 'nonce-eaDHPClCs73k19urySogiQ=='; 
      script-src https://cdn.amcharts.com 'self' 'nonce-eaDHPClCs73k19urySogiQ=='"
    />

    <script src="https://cdn.amcharts.com/lib/version/4.10.19/core.js"></script>
    <script src="https://cdn.amcharts.com/lib/version/4.10.19/charts.js"></script>
    <style nonce="eaDHPClCs73k19urySogiQ==">
      #chartdiv {
        width: 100%;
        height: 500px;
      }
    </style>
  </head>
  <body>
    <!-- HTML -->
    <div id="chartdiv"></div>

    <!-- Chart code -->
    <script nonce="eaDHPClCs73k19urySogiQ==">
      am4core.ready(function () {
        am4core.options.nonce = "eaDHPClCs73k19urySogiQ==";

        // Create chart instance
        let chart = am4core.create("chartdiv", am4charts.XYChart);

        chart.data = [
          {
            date: "2019-01-01",
            value: 1,
          },
          {
            date: "2019-02-01",
            value: 2,
          },
          {
            date: "2019-03-01",
            value: 3,
          },
        ];

        let dateAxis = chart.xAxes.push(new am4charts.DateAxis());
        let valueAxis = chart.yAxes.push(new am4charts.ValueAxis());

        let series = chart.series.push(new am4charts.LineSeries());
        series.dataFields.valueY = "value";
        series.dataFields.dateX = "date";

        /** Displaying "name" in the "tooltipText" will cause the CSP error **/
        series.name = "Name";
        series.tooltipText = `{name}: [bold]{valueY}[/]`;

        /** 
         Uncomment the line of code below and comment out the line above.
         You will see that the error will be gone.
        **/
        // series.tooltipText = `{valueY}`;

        chart.cursor = new am4charts.XYCursor();
      });
    </script>
  </body>
</html>
```

CodePen repro (note again that CodePen clouds the relevant CSP in Devtools so it's harder to see the error): https://codepen.io/mujahidfa/pen/NWpLrRV

## Environment
- amCharts `v4.10.19`

**Comment by martynasma** [comment]
I'm not entirely sure how to go about it. Even if we add `nonce` attribute to the `<tspan>` node (inline text formatting is applied via those nodes), it still triggers CSP "style-src" warning.

```SVG
<text x="0" y="21.359375" nonce="eaDHPClCs73k19urySogiQ==" dy="-5.767">
  <tspan>Name: </tspan>
  <tspan style="font-weight:bold" nonce="eaDHPClCs73k19urySogiQ==">1</tspan>
</text>
```

Any suggestions how we could work around that are welcome.

In the interim, I don't have any other suggestions other than using "unsafe-eval" in "style-src".

**Comment by mujahidfa** [comment]
@martynasma how about this? Add an inline `<style>` block (with a `nonce` attribute) containing the class style for the `<tspan>` :

```html
<text x="0" y="21.359375" dy="-5.767">
  <style nonce="eaDHPClCs73k19urySogiQ==">
    .cryptographic-hash-to-make-class-name-unique {
      font-weight: bold;
    }
  </style>
  <tspan>Name: </tspan>
  <tspan class="cryptographic-hash-to-make-class-name-unique">1</tspan>
</text>
```

> In the interim, I don't have any other suggestions other than using "unsafe-eval" in "style-src".

You mean "unsafe-inline", right?

**Comment by martynasma** [comment]
> You mean "unsafe-inline", right?

Yup. My bad.

> how about this? Add an inline <style> block (with a nonce attribute) containing the class style for the <tspan> :

That could work. Making a not to try it out.

**Comment by mujahidfa** [comment]
@martynasma another option is to use the `font-weight` attribute in `<tspan>` like so:

```html
<svg>
  <text x="0" y="21.359375" dy="-5.767">
    <tspan>Name:</tspan>
    <tspan font-weight="bold">1</tspan>
  </text>
</svg>
```

This way, we don't need to use nonce at all. However, this requires wrapping `<text>` around `<svg>` tags.

**Comment by martynasma** [comment]
OK, so the stint with `<style>` takes care of the issue. Thanks for the idea!

We'll implement it in the next version, and will let you know here, when it's out.

We've just released a new version, so not sure when the new version will be out, though. Might be 2-3 weeks.

**Comment by mujahidfa** [comment]
Awesome, glad to know it works! Would it be possible if you could release this version sooner than later? I actually really need this fix because this issue is actually blocking me from being able to use Amcharts in an internal app due to not passing the CSP security check. 

I'm a really big fan of Amcharts, but this issue is the only thing that is preventing me from using it in a prod app. Would really appreciate a speedier release this time around if it's not too much to ask 😊

**Comment by martynasma** [comment]
I'll see what I can do :) No promises, though. Any chance you could use "unsafe-inline" in the interim?

**Comment by mujahidfa** [comment]
Unfortunately no 😢 my org prohibits the usage of "unsafe-inline" so this is an actual deal breaker :( Was looking around for other chart libraries, and Amcharts seems to be the only few that supports nonce, which is one of the things that attracted me to Amcharts in the first place. If this issue is fixed, we'll be using Amcharts for sure!

**Comment by martynasma** [comment]
Gotcha. I'll see if we can make it happen for next week.

**Comment by mujahidfa** [comment]
Thank you, really appreciate it!!!

**Comment by martynasma** [comment]
Fixed in [4.10.20](https://github.com/amcharts/amcharts4/releases/tag/4.10.20).

## [4.10.20] - 2021-06-17

### Fixed
- In-line formatting of labels was triggering errors on pages with strict CSP rules.

[Full change log](https://github.com/amcharts/amcharts4/blob/master/dist/es2015/CHANGELOG.md).

[Download options](https://www.amcharts.com/download/).

Make sure you clear your browser cache after upgrading. And feel free to contact us again if you are still experiencing this issue.
